lib/util/flattenOperands.js

```diff
@@ -34,293 +35,80 @@ interested in how that works
 // 2+2+2, ie one + node that has three children.
 // Input: an expression tree
 // Output: the expression tree updated with flattened operations
-function flattenOperands(node) {
-  if (Node.Type.isConstant(node, true)) {
-    // the evaluate() changes unary minuses around constant nodes to constant nodes
-    // with negative values.
-    const constNode = Node.Creator.constant(evaluate(node));
-    if (node.changeGroup) {
-      constNode.changeGroup = node.changeGroup;
-    }
-    return constNode;
-  }
-  else if (Node.Type.isOperator(node)) {
-    if ('+-/*'.includes(node.op)) {
-      let parentOp;
-      if (node.op === '/') {
-        // Division is flattened in partner with multiplication. This means
-        // that after collecting the operands, they'll be children args of *
-        parentOp = '*';
-      }
-      else if (node.op === '-') {
-        // Subtraction is flattened in partner with addition, This means that
-        // after collecting the operands, they'll be children args of +
-        parentOp = '+';
+function flattenOperands(root) {
+  return replace(root, {
+    leave(node) {
+      if (query.isAdd(node)) {
+        return build.add(
+          ...node.args.reduce((accum, arg) => {
+            if (query.isAdd(arg)) {
+              return accum.concat(arg.args);
+            }
+            else {
+              return accum.concat(arg);
+            }
+          }, [])
+        );
       }
-      else {
-        parentOp = node.op;
+      else if (query.isMul(node)) {
+        // Rewrite explicit multiplication to be implicit if the factors look
+        // like those of a polynomial.  The first factor must be a number and
+        // the following factors must not be numbers.
+        // e.g. 2 * x^2 * y -> 2 x^2 y
+        if (isPolynomialTerm(node) && query.isNumber(node.args[0])
+            && node.args.slice(1).every(arg => !query.isNumber(arg))) {
+          const newNode = clone(node);
+          newNode.implicit = true;
+          return newNode;
+        }
+        // Rearrange multiplication of fractions involving variables in the
+        // the numerator.
+        // e.g. 2 x / 3 -> 2x / 3
+        else if (query.isNumber(node.args[0]) && query.isDiv(node.args[1])
```

Comment:
I've tried to set these things up so I can name things instead of just doing args[..].args[..]

i.e. something like

```
fraction = node.args[1]
numerator = fraction.args[0]
query.isIdentifier(numerator)
```

cause it can be pretty hard to read otherwise

I see how it could get messy here because it's all in the if statement, but I wonder if you could make a helper function? would that be harder to read? just throwing it out there